File: services/memory-store/src/memory/memory.service.ts
import { createHash } from 'node:crypto';

import {
  Injectable,
  Logger,
  NotFoundException,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { DatabaseService } from '../database/database.service';
import { RequestContextService } from '../observability/request-context.service';
import { CreateMemoryDto } from './dto/create-memory.dto';
import { QueryMemoryDto } from './dto/query-memory.dto';
import { SearchMemoryDto } from './dto/search-memory.dto';
import { UpdateMemoryDto } from './dto/update-memory.dto';

export interface MemoryRow {
  id: string;
  type: string;
  status: string;
  title: string;
  summary: string | null;
  content: string | null;
  content_hash: string;
  size: number;
  metadata: string;
  tags: string[];
  tenant_id: string;
  user_id: string | null;
  workspace_id: string | null;
  source_type: string | null;
  source_uri: string | null;
  source_label: string | null;
  created_at: Date;
  updated_at: Date;
}

@Injectable()
export class MemoryService {
  private readonly logger = new Logger(MemoryService.name);

  constructor(
    private readonly db: DatabaseService,
    private readonly config: ConfigService,
    private readonly requestContext: RequestContextService,
  ) {}

  async create(dto: CreateMemoryDto): Promise<MemoryRow> {
    const tenantId = this.requestContext.tenantId ?? 'default';
    const userId = this.requestContext.userId;

    const content = dto.content ?? '';
    const contentHash = createHash('sha256').update(content).digest('hex');
    const size = Buffer.byteLength(content, 'utf-8');

    const embedding = await this.computeEmbedding(content);

    const result = await this.db.query<MemoryRow>(
      `INSERT INTO memories (
        type, status, title, summary, content, content_hash, size,
        embedding, metadata, tags, tenant_id, user_id,
        source_type, source_uri, source_label
      ) VALUES ($1, 'INDEXED', $2, $3, $4, $5, $6, $7::vector, $8::jsonb, $9, $10, $11, $12, $13, $14)
      RETURNING *`,
      [
        dto.type,
        dto.title,
        dto.summary ?? null,
        dto.content ?? null,
        contentHash,
        size,
        `[${embedding.join(',')}]`,
        JSON.stringify(dto.metadata ?? {}),
        dto.tags ?? [],
        tenantId,
        userId ?? null,
        dto.sourceType ?? null,
        dto.sourceUri ?? null,
        dto.sourceLabel ?? null,
      ],
    );

    this.logger.log({ memoryId: result.rows[0]!.id }, 'Memory created');
    return result.rows[0]!;
  }

  async findAll(queryDto: QueryMemoryDto) {
    const page = queryDto.page ?? 1;
    const pageSize = queryDto.pageSize ?? 20;
    const { types, status, tags, dateFrom, dateTo, search } = queryDto;
    const offset = (page - 1) * pageSize;
    const tenantId = this.requestContext.tenantId ?? 'default';

    const conditions: string[] = ['m.tenant_id = $1'];
    const params: unknown[] = [tenantId];
    let paramIndex = 2;

    if (types && types.length > 0) {
      conditions.push(`m.type = ANY($${paramIndex}::text[])`);
      params.push(types);
      paramIndex++;
    }

    if (status) {
      conditions.push(`m.status = $${paramIndex}`);
      params.push(status);
      paramIndex++;
    }

    if (tags && tags.length > 0) {
      conditions.push(`m.tags && $${paramIndex}::text[]`);
      params.push(tags);
      paramIndex++;
    }

    if (dateFrom) {
      conditions.push(`m.created_at >= $${paramIndex}::timestamptz`);
      params.push(dateFrom);
      paramIndex++;
    }

    if (dateTo) {
      conditions.push(`m.created_at <= $${paramIndex}::timestamptz`);
      params.push(dateTo);
      paramIndex++;
    }

    if (search) {
      conditions.push(`m.title ILIKE $${paramIndex} OR m.summary ILIKE $${paramIndex} OR m.content ILIKE $${paramIndex}`);
      params.push(`%${search}%`);
      paramIndex++;
    }

    const whereClause = conditions.join(' AND ');

    const countResult = await this.db.query<{ count: string }>(
      `SELECT COUNT(*) as count FROM memories m WHERE ${whereClause}`,
      params,
    );

    const total = Number(countResult.rows[0]!.count);

    const dataResult = await this.db.query<MemoryRow>(
      `SELECT m.* FROM memories m WHERE ${whereClause}
       ORDER BY m.created_at DESC
       LIMIT $${paramIndex} OFFSET $${paramIndex + 1}`,
      [...params, pageSize, offset],
    );

    const totalPages = Math.ceil(total / pageSize);

    return {
      data: dataResult.rows,
      meta: {
        page,
        pageSize,
        total,
        totalPages,
        hasNext: page < totalPages,
        hasPrevious: page > 1,
      },
    };
  }

  async findById(id: string): Promise<MemoryRow> {
    const tenantId = this.requestContext.tenantId ?? 'default';

    const result = await this.db.query<MemoryRow>(
      `SELECT * FROM memories WHERE id = $1 AND tenant_id = $2`,
      [id, tenantId],
    );

    if (result.rows.length === 0) {
      throw new NotFoundException(`Memory with id "${id}" not found`);
    }

    return result.rows[0]!;
  }

  async update(id: string, dto: UpdateMemoryDto): Promise<MemoryRow> {
    const tenantId = this.requestContext.tenantId ?? 'default';

    const existing = await this.findById(id);

    const type = dto.type ?? existing.type;
    const title = dto.title ?? existing.title;
    const summary = dto.summary !== undefined ? dto.summary : existing.summary;
    const content = dto.content !== undefined ? dto.content : existing.content;
    const sourceType = dto.sourceType !== undefined ? dto.sourceType : existing.source_type;
    const sourceUri = dto.sourceUri !== undefined ? dto.sourceUri : existing.source_uri;
    const sourceLabel = dto.sourceLabel !== undefined ? dto.sourceLabel : existing.source_label;
    const metadata = dto.metadata !== undefined ? dto.metadata : JSON.parse(existing.metadata);
    const tags = dto.tags !== undefined ? dto.tags : existing.tags;

    const body = content ?? '';
    const contentHash = createHash('sha256').update(body).digest('hex');
    const size = Buffer.byteLength(body, 'utf-8');

    const contentChanged = content !== existing.content;
    let embedding: number[] | null = null;

    if (contentChanged && content) {
      embedding = await this.computeEmbedding(content);
    }

    const result = await this.db.query<MemoryRow>(
      `UPDATE memories SET
        type = $1, title = $2, summary = $3, content = $4,
        content_hash = $5, size = $6, metadata = $7::jsonb, tags = $8,
        source_type = $9, source_uri = $10, source_label = $11,
        embedding = COALESCE($12::vector, embedding),
        updated_at = NOW()
      WHERE id = $13 AND tenant_id = $14
      RETURNING *`,
      [
        type,
        title,
        summary,
        content ?? null,
        contentHash,
        size,
        JSON.stringify(metadata),
        tags,
        sourceType,
        sourceUri,
        sourceLabel,
        embedding ? `[${embedding.join(',')}]` : null,
        id,
        tenantId,
      ],
    );

    this.logger.log({ memoryId: id }, 'Memory updated');
    return result.rows[0]!;
  }

  async softDelete(id: string): Promise<void> {
    const tenantId = this.requestContext.tenantId ?? 'default';

    const result = await this.db.query(
      `UPDATE memories SET status = 'DELETED', updated_at = NOW()
       WHERE id = $1 AND tenant_id = $2`,
      [id, tenantId],
    );

    if (result.rowCount === 0) {
      throw new NotFoundException(`Memory with id "${id}" not found`);
    }

    this.logger.log({ memoryId: id }, 'Memory soft-deleted');
  }

  async search(dto: SearchMemoryDto): Promise<{ data: MemoryRow[]; meta: { total: number } }> {
    const tenantId = this.requestContext.tenantId ?? 'default';

    const embedding = await this.computeEmbedding(dto.query);
    const embeddingStr = `[${embedding.join(',')}]`;

    const conditions: string[] = ['m.tenant_id = $1', "m.status != 'DELETED'"];
    const params: unknown[] = [tenantId];
    let paramIndex = 2;

    if (dto.types && dto.types.length > 0) {
      conditions.push(`m.type = ANY($${paramIndex}::text[])`);
      params.push(dto.types);
      paramIndex++;
    }

    if (dto.tags && dto.tags.length > 0) {
      conditions.push(`m.tags && $${paramIndex}::text[]`);
      params.push(dto.tags);
      paramIndex++;
    }

    const whereClause = conditions.join(' AND ');

    const limit = dto.limit ?? 10;

    const result = await this.db.query<MemoryRow & { distance: number }>(
      `SELECT m.*, m.embedding <=> $${paramIndex}::vector AS distance
       FROM memories m
       WHERE ${whereClause}
         AND m.embedding IS NOT NULL
       ORDER BY distance ASC
       LIMIT $${paramIndex + 1}`,
      [...params, embeddingStr, limit],
    );

    const minScore = dto.minScore ?? 0;
    const filtered = result.rows.filter((row) => {
      const score = 1 / (1 + row.distance);
      return score >= minScore;
    });

    return {
      data: filtered,
      meta: { total: filtered.length },
    };
  }

  private async computeEmbedding(text: string): Promise<number[]> {
    if (!text || text.trim().length === 0) {
      return new Array(1536).fill(0);
    }

    const aiUrl = this.config.get<string>('ai.serviceUrl') as string;
    const endpoint = this.config.get<string>('ai.embeddingEndpoint') as string;

    try {
      const response = await fetch(`${aiUrl}${endpoint}`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ text: text.slice(0, 8192) }),
      });

      if (!response.ok) {
        throw new Error(`AI service returned ${response.status}: ${await response.text()}`);
      }

      const body = (await response.json()) as { embedding: number[] } | { data: { embedding: number[] }[] };
      const embedding = 'data' in body ? (body.data[0]?.embedding ?? null) : body.embedding;

      if (!embedding || !Array.isArray(embedding)) {
        throw new Error('Invalid embedding response from AI service');
      }

      return embedding;
    } catch (err) {
      this.logger.warn({ err }, 'Embedding computation failed, using zero vector');
      return new Array(1536).fill(0);
    }
  }
}

File: apps/ai-service/src/app/services/memory_service.py
import uuid
from typing import Any

from sqlalchemy import select, func, or_, delete
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from ..models.memory import Memory
from ..schemas.memory import MemoryCreate, MemoryUpdate, MemoryQuery, MemorySearch
from .llm_service import llm_service, LLMProviderError


class MemoryService:
    async def create_memory(
        self, db: AsyncSession, dto: MemoryCreate, tenant_id: str | None, user_id: str | None
    ) -> Memory:
        content_for_embedding = dto.content or dto.title or dto.summary or ""
        embedding = None
        if content_for_embedding.strip():
            try:
                embedding = await llm_service.generate_embedding(content_for_embedding)
            except LLMProviderError:
                embedding = None

        memory = Memory(
            id=uuid.uuid4(),
            type=dto.type,
            title=dto.title,
            summary=dto.summary,
            content=dto.content,
            content_hash=llm_service.compute_content_hash(content_for_embedding) if content_for_embedding else None,
            size=len(content_for_embedding) if content_for_embedding else 0,
            embedding=embedding,
            metadata=dto.metadata or {},
            tags=dto.tags,
            tenant_id=tenant_id,
            user_id=user_id,
            workspace_id=dto.workspace_id,
            source_type=dto.source_type,
            source_uri=dto.source_uri,
            source_label=dto.source_label,
            connector_id=dto.connector_id,
        )
        db.add(memory)
        await db.flush()
        return memory

    async def list_memories(self, db: AsyncSession, query: MemoryQuery, tenant_id: str | None) -> tuple[list[Memory], int]:
        stmt = select(Memory)
        count_stmt = select(func.count(Memory.id))

        conditions = [Memory.status == (query.status or "active")]
        if query.type:
            conditions.append(Memory.type == query.type)
        if tenant_id:
            conditions.append(Memory.tenant_id == tenant_id)
        if query.tags:
            conditions.append(Memory.tags.overlap(query.tags))

        stmt = stmt.where(*conditions).order_by(Memory.created_at.desc())
        count_stmt = count_stmt.where(*conditions)

        offset = (query.page - 1) * query.page_size
        stmt = stmt.offset(offset).limit(query.page_size)

        total_result = await db.execute(count_stmt)
        total = total_result.scalar_one()

        result = await db.execute(stmt)
        memories = list(result.scalars().all())

        return memories, total

    async def get_memory(self, db: AsyncSession, memory_id: uuid.UUID, tenant_id: str | None) -> Memory | None:
        stmt = select(Memory).where(Memory.id == memory_id)
        if tenant_id:
            stmt = stmt.where(Memory.tenant_id == tenant_id)
        result = await db.execute(stmt)
        return result.scalar_one_or_none()

    async def update_memory(self, db: AsyncSession, memory_id: uuid.UUID, dto: MemoryUpdate, tenant_id: str | None) -> Memory | None:
        memory = await self.get_memory(db, memory_id, tenant_id)
        if not memory:
            return None

        update_data = dto.model_dump(exclude_unset=True)

        if "content" in update_data and update_data["content"] is not None:
            content_for_embedding = update_data.get("content") or memory.content or ""
            if content_for_embedding.strip():
                try:
                    update_data["embedding"] = await llm_service.generate_embedding(content_for_embedding)
                except LLMProviderError:
                    pass
                update_data["content_hash"] = llm_service.compute_content_hash(content_for_embedding)
                update_data["size"] = len(content_for_embedding)

        for key, value in update_data.items():
            setattr(memory, key, value)

        await db.flush()
        return memory

    async def delete_memory(self, db: AsyncSession, memory_id: uuid.UUID, tenant_id: str | None) -> bool:
        memory = await self.get_memory(db, memory_id, tenant_id)
        if not memory:
            return False
        memory.status = "deleted"
        await db.flush()
        return True

    async def search_memories(
        self, db: AsyncSession, dto: MemorySearch, tenant_id: str | None
    ) -> list[tuple[Memory, float]]:
        content_for_embedding = dto.query
        query_embedding = await llm_service.generate_embedding(content_for_embedding)

        stmt = select(Memory, func.cosine_distance(Memory.embedding, query_embedding).label("distance"))
        conditions = [Memory.status == "active", Memory.embedding.isnot(None)]
        if tenant_id:
            conditions.append(Memory.tenant_id == tenant_id)
        if dto.type:
            conditions.append(Memory.type == dto.type)
        if dto.tags:
            conditions.append(Memory.tags.overlap(dto.tags))

        stmt = stmt.where(*conditions)
        if dto.threshold is not None:
            stmt = stmt.where(func.cosine_distance(Memory.embedding, query_embedding) <= (1.0 - dto.threshold))
        stmt = stmt.order_by(func.cosine_distance(Memory.embedding, query_embedding)).limit(dto.top_k)

        result = await db.execute(stmt)
        rows = result.all()
        return [(row[0], float(1.0 - row[1])) for row in rows]


memory_service = MemoryService()
